Kernel/Memory/MemoryManager.cpp

```diff
@@ -278,152 +280,370 @@ UNMAP_AFTER_INIT void MemoryManager::parse_memory_map()
-            if (diff != 0) {
-                dmesgln("MM: Got an unaligned physical_region from the bootloader; correcting {:p} by {} bytes", address, diff);
-                diff = PAGE_SIZE - diff;
-                address += diff;
-                length -= diff;
-            }
-            if ((length % PAGE_SIZE) != 0) {
-                dmesgln("MM: Got an unaligned physical_region from the bootloader; correcting length {} by {} bytes", length, length % PAGE_SIZE);
-                length -= length % PAGE_SIZE;
-            }
-            if (length < PAGE_SIZE) {
-                dmesgln("MM: Memory physical_region from bootloader is too small; we want >= {} bytes, but got {} bytes", PAGE_SIZE, length);
-                continue;
-            }
+        for (auto& region : global_data.physical_regions) {
+            dmesgln("MM: User physical region: {} - {} (size {:#x})", region->lower(), region->upper().offset(-1), PAGE_SIZE * region->size());
+            region->initialize_zones();
+        }
+    });
+}
+
+UNMAP_AFTER_INIT void MemoryManager::parse_memory_map_fdt(MemoryManager::GlobalData& global_data, PhysicalPtr fdt_addr)
+{
+    auto fdt_header = *reinterpret_cast<DeviceTree::FlattenedDeviceTreeHeader const*>(quickmap_page(PhysicalAddress { fdt_addr }));
+    unquickmap_page();
+    // FIXME: Actually map this
+    u8* fdt_buffer_pointer = (u8*)__builtin_alloca(MUST(page_round_up(fdt_header.totalsize)));
+
+    for (FlatPtr offset = 0; offset < MUST(page_round_up(fdt_header.totalsize)); offset += PAGE_SIZE) {
+        auto* src = quickmap_page(PhysicalAddress { fdt_addr + offset });
+        __builtin_memcpy(&fdt_buffer_pointer[offset], src, PAGE_SIZE);
+        unquickmap_page();
+    }
+    auto fdt_buffer = ReadonlyBytes(fdt_buffer_pointer, fdt_header.totalsize);
+    // FIXME: Parse the MemoryReservationBlock
 
-            // FIXME: This might have a nicer solution than slicing the ranges apart,
-            //        to just put them back together when we dont find a used range in them
-            for (PhysicalSize page_base = address; page_base <= (address + length); page_base += PAGE_SIZE) {
-                auto addr = PhysicalAddress(page_base);
-
-                // Skip used memory ranges.
-                bool should_skip = false;
-                for (auto& used_range : global_data.used_memory_ranges) {
-                    if (addr.get() >= used_range.start.get() && addr.get() <= used_range.end.get()) {
-                        should_skip = true;
-                        page_base = used_range.end.get();
+    enum class State {
+        Root,
+        InReservedMemory,
+        InReservedMemoryChild,
+
+        InMemory
+    };
+
+    struct {
+        u32 depth = 0;
+        State state = State::Root;
+        u64 start = 0;
+        u64 size = 0;
+        u32 address_cells = 0;
+        u32 size_cells = 0;
+    } state;
+
+    // FIXME: Maybe readout the MMIO ranges of the devices in here
```

Comment:
Probably not necessary until after we've unflattened the device tree and moved it to permanent memory.